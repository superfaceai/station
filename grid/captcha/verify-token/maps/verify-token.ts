/* eslint-disable jest/no-export */

import { SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

export const verifyTokenTest = (provider: string, validResponse: string) => {
  describe(`captcha/verify-token/${provider}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = buildSuperfaceTest({
        profile: 'captcha/verify-token',
        provider,
      });
    });

    describe('ValidateCaptchaResponse', () => {
      it('should map error', async () => {
        const result = await superface.run({
          useCase: 'ValidateCaptchaResponse',
          input: {
            response: 'invalid response',
            idempotencyKey: '123e4567-e89b-12d3-a456-426614174000',
          },
        });

        expect(() => result.unwrap()).toThrow();
        expect(result).toMatchSnapshot();
      });

      it('should perform successfully', async () => {
        const result = await superface.run({
          useCase: 'ValidateCaptchaResponse',
          input: {
            response: validResponse,
            idempotencyKey: '123e4567-e89b-12d3-a456-426614174000',
          },
        });

        console.debug(result);

        expect(() => result.unwrap()).not.toThrow();
        expect(result).toMatchSnapshot();
      });
    });
  });
};
