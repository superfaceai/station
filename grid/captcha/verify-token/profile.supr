name = "captcha/verify-token"
version = "1.0.0"

"Validates the CAPTCHA provider response from the website's backend."
usecase ValidateCaptchaResponse idempotent  {
  input {
    "The user response token to be verified, provided by CAPTCHA provider's client-side integration on your site."
    response! string!
  
    "A key to be used for idempotency in case of retrying failed requests."
    idempotencyKey string!

    "The user's IP address."
    remoteIp string!

    "The site key you expect to see."
    siteKey string!
  }

  result {
    "Timestamp of the challenge."
    challengeTimeStamp! string!
  
    "Hostname of the site where the CAPTCHA was solved."
    hostname! string!
  }

  error {
    "List of error codes related to the failed token validation."
    errorCodes! [string!]!
  
    "Timestamp of the challenge."
    challengeTimeStamp! string!
  
    "Hostname of the site where the CAPTCHA was solved."
    hostname! string!
  
    "The rate limit error, if applicable."
    rateLimit number
  }

  example SuccessExample {
    input {
      response = 'valid token',
      idempotencyKey = '123e4567-e89b-12d3-a456-426614174000',
    }
    result {
      challengeTimeStamp = '2022-02-28T15:14:30.096Z',
      hostname = 'example.com'
    }
  }

  example ErrorExample {
    input {
      response = 'invalid token',
      idempotencyKey = '123e4567-e89b-12d3-a456-426614174000',
    }
    error {
      errorCodes = ["invalid-input-response"]
      challengeTimeStamp = '2022-02-28T15:14:30.096Z',
      hostname = 'example.com'
    }
  }
}
