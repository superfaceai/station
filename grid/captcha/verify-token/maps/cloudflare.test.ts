import { verifyTokenTest } from './verify-token';

verifyTokenTest(
  'cloudflare',
  '0.FmQzukXdsyKTEB_VI0ZrKwmWIEl06A-UyzAkI1hURUU5z5uNXJWC3cEZjQG_39mS7jS0mJp910rAOe9W5g9MseSNnaHZIj4FYGFBzUxqxz_P5IGmgK-VyMUM-jHvA1YxMltTqjjDtyVOgrKcWberk5Ui-hAb9mHfdRuPiUApgsm_bd5VhUASssukyVcoqhHtSGl9ksOfX5teD9gwcuSGlxvD0W_RAhPSznyk-zRO_di5LTGd16MKOrxQFam950zLnlOL3-Y_YyLoQRogz9rPfP-hDQGrBkjZQotZ0zbqz6hVNVNqJ5NZ3eS49_fSFn7SpQ8RULVg4u9GSlnxiNnMOKev5K5SFRa7u5Ww1zDtlit4e2ALLrxu2-_Dsk4Jh31D.9zv2kt3VmsxZ2HUBOZEmhg.21b1ab766834a5662b6391d6979e92fb71ae10a8734d2279fb39a7ad413151db'
);
