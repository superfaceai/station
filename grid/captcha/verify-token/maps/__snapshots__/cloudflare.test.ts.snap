// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captcha/verify-token/cloudflare ValidateCaptchaResponse should map error 1`] = `
Err {
  "error": "MappedHTTPError: Expected HTTP error
Properties: {
  \\"errorCodes\\": [
    \\"invalid-input-response\\"
  ],
  \\"detail\\": \\"Challenge verification failed with error code(s) invalid-input-response.\\"
}
AST Path: definitions[0].statements[0].responseHandlers[0].statements[2]
Original Map Location: Line 24, column 7",
}
`;

exports[`captcha/verify-token/cloudflare ValidateCaptchaResponse should perform successfully 1`] = `
Ok {
  "value": Object {
    "challengeTimeStamp": "2023-06-06T10:58:18.833Z",
    "hostname": "localhost",
  },
}
`;
